File: 22-generate-parentheses/22-generate-parentheses.scala
import scala.collection.mutable._

object Solution {
    def generateParenthesis(n: Int): List[String] = {
        val res: ListBuffer[String] = ListBuffer()

        if (n == 0) {
            res += ""
        } else {
            for (i <- 0 to n - 1) {
                for(left <- generateParenthesis(i)) {
                    for(right <- generateParenthesis(n - i - 1)) {
                        res += s"${left}(${right})"
                    }
                }
            }
        }
        
        res.toList
    }
}

File: 22-generate-parentheses/22-generate-parentheses.java
class Solution {
    Map<Integer, List<String>> memo = new HashMap<>();
    
    public List<String> generateParenthesis(int n) {
        if (n == 0) return List.of("");
        
        if (memo.containsKey(n)) return memo.get(n);
        
        List<String> res = new ArrayList<>();
        
        for(int i = 0; i < n; i++) {
            for(String s1 : generateParenthesis(i)) {
                for(String s2 : generateParenthesis(n - i - 1)) {
                    res.add(s1 + "(" + s2 + ")");
                }
            }
        }
        
        memo.put(n, res);
        
        return res;
    }
}
